src/core/lib/attribute-transition-manager.js

```diff
@@ -0,0 +1,373 @@
+      const transition = this.attributeTransitions[attributeName];
+
+      buffers[transition.bufferIndex] = transition.buffer;
+
+      let time = 1;
+      if (transition.state === TRANSITION_STATE.PENDING) {
+        transition.startTime = currentTime;
+        transition.state = TRANSITION_STATE.STARTED;
+        transition.onStart(transition);
+      }
+
+      if (transition.state === TRANSITION_STATE.STARTED) {
+        time = (currentTime - transition.startTime) / transition.duration;
+        if (time >= 1) {
+          time = 1;
+          transition.state = TRANSITION_STATE.ENDED;
+          transition.onEnd(transition);
+        }
+        needsRedraw = true;
+      }
+
+      uniforms[`${transition.name}Time`] = transition.easing(time);
+    }
+
+    if (needsRedraw) {
+      this._runTransformFeedback({uniforms, buffers});
+    }
+
+    return needsRedraw;
+  }
+  /* eslint-enable max-statements */
+
+  /* Private methods */
+
+  // check the latest attributes for updates.
+  // Returns `true` if attributes have been added/removed.
+  _updateAttributes(attributes) {
+    let needsNewModel = false;
+
+    for (const attributeName in attributes) {
+      const attribute = attributes[attributeName];
+      const settings = this._getTransitionSettings(attribute);
+
+      if (settings) {
+        let hasChanged;
+        let transition = this.attributeTransitions[attributeName];
+        if (transition) {
+          hasChanged = attribute.changed;
+        } else {
+          // New animated attributes have been added
+          transition = {name: attributeName, attribute};
+          this.attributeTransitions[attributeName] = transition;
+          hasChanged = true;
+          needsNewModel = true;
+        }
+
+        if (hasChanged) {
+          this._triggerTransition(transition, settings);
+          this._updateModel(attributeName, transition);
+          this.needsRedraw = true;
+        }
+      }
+    }
+
+    for (const attributeName in this.attributeTransitions) {
+      const attribute = attributes[attributeName];
+
+      if (!attribute || !attribute.transition) {
+        // Animated attribute has been removed
+        delete this.attributeTransitions[attributeName];
+        needsNewModel = true;
+      }
+    }
+
+    return needsNewModel;
+  }
+
+  // Redraw the transform feedback
+  _runTransformFeedback({uniforms, buffers}) {
+    const {model, transformFeedback} = this;
+
+    // Cannot bind to transform feedback if it's already used as attribute
+    Object.values(buffers).forEach(buffer => buffer.unbind());
+
+    model.program.use();
+    transformFeedback.bindBuffers(buffers, {clear: true});
+    transformFeedback.begin(GL.POINTS);
+
+    const parameters = {
+      [GL.RASTERIZER_DISCARD]: true
+    };
+
+    model.draw({uniforms, parameters});
+
+    transformFeedback.end();
+  }
+
+  // Create a model for the transform feedback
+  _createModel() {
```

Comment:
Could this be broken into a separate file and be made more reusable? Could be a "seed" for the GPGPU lib.
  
Regardless, would be nice to get all the shader and webgl logic out of this long file.